cloud/scope/powervs_machine_test.go

```diff
@@ -122,74 +133,1186 @@ func newDHCPServer(serverID, networkID string) models.DHCPServers {
+			cosInstanceName := fmt.Sprintf("%s-%s", scope.IBMPowerVSCluster.GetName(), "cosinstance")
+			mockResourceController.EXPECT().GetInstanceByName(cosInstanceName, resourcecontroller.CosResourceID, resourcecontroller.CosResourcePlanID).Return(nil, nil)
+			scope.ResourceClient = mockResourceController
+			result, err := scope.createCOSClient()
+			g.Expect(result).To(BeNil())
+			g.Expect(err).To(BeNil())
+		})
+
+		t.Run("COS service instance is not in active state", func(t *testing.T) {
+			g := NewWithT(t)
+			setup(t)
+			t.Cleanup(teardown)
+			scope := setupPowerVSMachineScope(clusterName, machineName, core.StringPtr(pvsImage), core.StringPtr(pvsNetwork), true, mockpowervs)
+			serviceInstance := new(resourcecontrollerv2.ResourceInstance)
+			serviceInstance.State = core.StringPtr(string(infrav1beta2.ServiceInstanceStateProvisioning))
+			cosInstanceName := fmt.Sprintf("%s-%s", scope.IBMPowerVSCluster.GetName(), "cosinstance")
+			mockResourceController.EXPECT().GetInstanceByName(cosInstanceName, resourcecontroller.CosResourceID, resourcecontroller.CosResourcePlanID).Return(serviceInstance, nil)
+			scope.ResourceClient = mockResourceController
+			result, err := scope.createCOSClient()
+			expectedError := fmt.Sprintf("COS service instance is not in active state, current state: %s", infrav1beta2.ServiceInstanceStateProvisioning)
+			g.Expect(result).To(BeNil())
+			g.Expect(err.Error()).To(ContainSubstring(expectedError))
+		})
+
+		t.Run("Failed to determine COS bucket region", func(t *testing.T) {
+			g := NewWithT(t)
+			setup(t)
+			t.Cleanup(teardown)
+			scope := setupPowerVSMachineScope(clusterName, machineName, core.StringPtr(pvsImage), core.StringPtr(pvsNetwork), true, mockpowervs)
+			serviceInstance := new(resourcecontrollerv2.ResourceInstance)
+			serviceInstance.State = core.StringPtr(string(infrav1beta2.ServiceInstanceStateActive))
+			scope.SetRegion(region)
+			cosInstanceName := fmt.Sprintf("%s-%s", scope.IBMPowerVSCluster.GetName(), "cosinstance")
+			mockResourceController.EXPECT().GetInstanceByName(cosInstanceName, resourcecontroller.CosResourceID, resourcecontroller.CosResourcePlanID).Return(serviceInstance, nil)
+			scope.ResourceClient = mockResourceController
+			result, err := scope.createCOSClient()
+			expectedError := "failed to determine COS bucket region, both bucket region and VPC region not set"
+			g.Expect(result).To(BeNil())
+			g.Expect(err.Error()).To(ContainSubstring(expectedError))
+		})
+		t.Run("Creates COS client successfully", func(t *testing.T) {
+			g := NewWithT(t)
+			setup(t)
+			t.Cleanup(teardown)
+			scope := setupPowerVSMachineScope(clusterName, machineName, core.StringPtr(pvsImage), core.StringPtr(pvsNetwork), true, mockpowervs)
+			serviceInstance := new(resourcecontrollerv2.ResourceInstance)
+			serviceInstance.State = core.StringPtr(string(infrav1beta2.ServiceInstanceStateActive))
+			guid := "foo-guid"
+			serviceInstance.GUID = &guid
```

Comment:
```suggestion
			serviceInstance := resourcecontrollerv2.ResourceInstance{
		               State: ptr.To(string(infrav1beta2.ServiceInstanceStateActive)),
		               GUID:  ptr.To("foo-guid"),
	}
```
How about we simplify wherever applicable?